File: src/main/scala-2.11/com/lei/leetcode/P328/Solution.java
package com.lei.leetcode.P328;

import com.lei.leetcode.ListNode;

/**
 * Created by Lei on 2017/7/7.
 */
public class Solution {
    public static ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode odd = new ListNode(-1);
        ListNode even = new ListNode(-1);
        ListNode op = odd, ep = even;
        int index = 1;
        while (head != null) {
            if (index % 2 == 1) {
                op.next = head;
                op = op.next;
            } else {
                ep.next = head;
                ep = ep.next;
            }
            index++;
            head = head.next;
        }
        op.next = null;
        ep.next = null;

        op = odd;
        while (op.next != null) {
            op = op.next;
        }
        op.next = even.next;
        return odd.next;


    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);
        ListNode p = oddEvenList(head);
        while (p != null) {
            System.out.println(p.val);
            p = p.next;
        }
    }
}

File: src/main/scala-2.11/com/lei/leetcode/P328/Solution.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return None

        op_dummy = op = ListNode(0)
        ep_dummy = ep = ListNode(0)
        while head:
            op.next = head
            ep.next = head.next
            ep = ep.next
            op = op.next
            head = head.next.next if ep else None
        op.next = ep_dummy.next
        return op_dummy.next


# 1 2 3 4 5 6
# 1 2 3 4 5

if __name__ == '__main__':
    s = Solution()
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    print s.oddEvenList(head)
